File: band/band-schedule/src/common/protocol.js
/**
 * 下发端(电脑端 / 手机端) <-> 手环端(RPK) 的互联协议。
 *
 * 手环端 interconnect 的 onmessage 收到的是下发端 sendMessage 的原始字符串，
 * 这里统一封装成 JSON 信封：
 *   { "t": "<type>", "d": <payload> }
 *
 * 方向：下发端 -> 手环
 *   hello     下发端握手，告知自己已就绪
 *   schedule  整份课表（课表较小时直接一帧发完）
 *   part      课表分片，d = { seq, total, data }
 *   config    下发时间表/开学日期等配置
 *
 * 方向：手环 -> 下发端
 *   ready     手环端就绪，请求同步
 *   request   用户在手环上主动点了「同步」
 *   ack       收到数据后的确认，d = { ok, courses, parts }
 *
 * 为什么要分片：发给手环的消息要经过系统服务中转，
 * 官方示例里的单帧缓冲是 1KB，一整份课表（几十门课）远超这个量级，
 * 一次发完可能被截断或直接失败。所以下发端按 300 个字符切帧
 * （中文 UTF-8 一字最多 3 字节，单帧约 900 字节，稳在 1KB 以内），
 * 手环端在 [createAssembler] 里拼回来。
 */

export const MSG = {
  HELLO: 'hello',
  SCHEDULE: 'schedule',
  PART: 'part',
  CONFIG: 'config',
  READY: 'ready',
  REQUEST: 'request',
  ACK: 'ack',
}

/** 分片超过这个时间还没收齐就丢弃，避免和下一轮同步混在一起。 */
const ASSEMBLE_TIMEOUT_MS = 60 * 1000

export function encode(type, payload) {
  return JSON.stringify({ t: type, d: payload === undefined ? null : payload })
}

/** 把字节按 UTF-8 解成字符串。
 *
 * 手环这边的 onmessage 给到的 `data` 可能是字符串、也可能是字节数组
 * （Uint8Array / ArrayBuffer / 普通数字数组）。下发端是按 UTF-8 编码发的，
 * 所以字节必须按 UTF-8 解 —— 一个字节当成一个字符去拼是错的：中文会变成
 * 乱码，JSON.parse 跟着失败，消息就被静默丢掉了。
 */
export function bytesToText(bytes) {
  if (bytes === null || bytes === undefined) return ''
  const list =
    bytes instanceof ArrayBuffer
      ? Array.from(new Uint8Array(bytes))
      : Array.isArray(bytes) || (bytes && typeof bytes.length === 'number')
        ? Array.prototype.slice.call(bytes)
        : null
  if (!list) return String(bytes)

  let out = ''
  let i = 0
  while (i < list.length) {
    const b = list[i] & 0xff
    let code = 0
    let extra = 0
    if (b < 0x80) {
      code = b
      extra = 0
    } else if ((b & 0xe0) === 0xc0) {
      code = b & 0x1f
      extra = 1
    } else if ((b & 0xf0) === 0xe0) {
      code = b & 0x0f
      extra = 2
    } else if ((b & 0xf8) === 0xf0) {
      code = b & 0x07
      extra = 3
    } else {
      // 非法起始字节：跳过，别让整条消息烂掉
      i += 1
      continue
    }
    if (i + extra >= list.length) break
    for (let k = 1; k <= extra; k += 1) {
      code = (code << 6) | ((list[i + k] & 0xff) & 0x3f)
    }
    i += extra + 1
    if (code > 0xffff) {
      // 需要一对代理
      const v = code - 0x10000
      out += String.fromCharCode(0xd800 + (v >> 10), 0xdc00 + (v & 0x3ff))
    } else {
      out += String.fromCharCode(code)
    }
  }
  return out
}

export function decode(raw) {
  if (raw === null || raw === undefined) return null

  // 1) 已经是结构化对象（固件直接给 {t, d}）—— 直接用
  if (typeof raw === 'object' && typeof raw.t === 'string') return raw

  // 2) 剥掉 { data: ... } / { payload: ... } 这层包装；没有包装就用 raw 本身
  let text = raw
  if (typeof text === 'object') {
    const inner = text.data !== undefined ? text.data : text.payload
    text = inner === undefined || inner === null ? raw : inner
    // 里面的东西本身还是结构化对象
    if (typeof text === 'object' && typeof text.t === 'string') return text
  }

  // 3) 字节（Uint8Array / ArrayBuffer / 数字数组）→ 按 UTF-8 解成字符串
  if (typeof text !== 'string') {
    text = bytesToText(text)
  }

  // 4) 字符串 → JSON
  //
  // ⚠️ 顺序很重要：有的固件把 UTF-8 字节按"每字节一字符"塞进字符串
  // （Latin-1 那种），中文会变成一串乱码字符。这种串 **JSON.parse 是能成功的**
  // （键还是 ASCII，只是值全是乱码），所以不能等解析失败才补救 ——
  // 必须先识别、先修复，否则拿到的是"能解析但内容是乱码"的结果。
  const bytes = bytePerCharBytes(text)
  if (bytes) {
    const repaired = tryParse(bytesToText(bytes))
    if (repaired) {
      console.log('[protocol] decode: payload was byte-per-char, fixed as UTF-8')
      return repaired
    }
  }

  const parsed = tryParse(text)
  if (parsed) return parsed
  return null
}

/**
 * 如果这个字符串是"每字节一字符"的畸形串，返回对应的字节数组；否则返回 null。
 *
 * 判据：所有字符码都 ≤ 0xFF（真文本里的中文会 > 0xFF），且**至少有一个**
 * 高位字符（纯 ASCII 的串不需要修复，直接解析就好）。
 */
function bytePerCharBytes(text) {
  let hasHighByte = false
  const bytes = []
  for (let i = 0; i < text.length; i += 1) {
    const c = text.charCodeAt(i)
    if (c > 0xff) return null
    if (c >= 0x80) hasHighByte = true
    bytes.push(c)
  }
  return hasHighByte ? bytes : null
}

/** 解析失败返回 null（不抛）。 */
function tryParse(text) {
  try {
    const parsed = JSON.parse(text)
    return parsed && parsed.t ? parsed : null
  } catch (e) {
    return null
  }
}

/**
 * 分片重组器。下发端可能把课表切成多帧发送，这里按 seq 收齐后再交给业务。
 *
 * accept(msg) 的返回值：
 *   { status: 'ignore' }                    不是课表相关的消息
 *   { status: 'progress', got, total }      还在收，已收 got 片
 *   { status: 'done', value }               收齐，value 是课表对象
 *   { status: 'error', reason }             出错，需要重新同步
 */
export function createAssembler() {
  let total = 0
  let chunks = []
  let got = 0
  let startedAt = 0

  function reset() {
    total = 0
    chunks = []
    got = 0
    startedAt = 0
  }

  return {
    reset,

    accept(msg) {
      if (!msg) return { status: 'ignore' }

      // 整包直发：小课表走这条，收到即可用
      if (msg.t === MSG.SCHEDULE) {
        reset()
        return { status: 'done', value: msg.d }
      }

      if (msg.t !== MSG.PART) return { status: 'ignore' }

      const d = msg.d || {}
      const seq = Number(d.seq)
      const count = Number(d.total)
      if (!isFinite(seq) || !isFinite(count) || count <= 0 || seq < 0 || seq >= count) {
        return { status: 'error', reason: 'bad part header' }
      }
      if (typeof d.data !== 'string') {
        return { status: 'error', reason: 'bad part body' }
      }

      const now = Date.now()
      const stale = startedAt > 0 && now - startedAt > ASSEMBLE_TIMEOUT_MS
      // 新的一轮同步（片数变了或超时了）就丢弃上一轮的残留
      if (total !== count || stale) {
        reset()
        total = count
        startedAt = now
      }
      if (startedAt === 0) startedAt = now

      if (chunks[seq] === undefined) {
        chunks[seq] = d.data
        got++
      }

      if (got < total) return { status: 'progress', got: got, total: total }

      let text = ''
      for (let i = 0; i < total; i++) {
        if (chunks[i] === undefined) return { status: 'error', reason: 'missing part ' + i }
        text += chunks[i]
      }
      reset()

      try {
        return { status: 'done', value: JSON.parse(text) }
      } catch (e) {
        return { status: 'error', reason: 'json: ' + e.message }
      }
    },
  }
}

File: pc/lib/src/band/protocol.dart
/// 与手环端快应用之间的互联协议。
///
/// ⚠️ 必须与手环端 `band/band-schedule/src/common/protocol.js` 完全一致。
/// 两端对不上的表现是"发出去没反应"或"手环显示数据异常"。
///
/// 信封：`{"t": "<type>", "d": <payload>}`
///
/// 方向：
///   电脑端 → 手环   hello      握手
///                    schedule   课表较小，一帧发完
///                    part       课表分片 {seq, total, data}
///   手环 → 电脑端   ready      手环就绪，请求同步
///                    request    用户在手环上点了同步
///                    ack        收到数据后的确认 {ok, courses} 或 {ok:false, error}
///
/// ### 为什么要分帧
/// 发给手环的消息经系统服务中转，官方示例里的单帧缓冲是 1KB，
/// 而一份真实课表序列化后有好几 KB。所以按 [chunkChars] 个字符切片
/// （中文一字最多 3 字节，300 字符 ≈ 900 字节，稳在 1KB 以内），
/// 手环端在 `createAssembler` 里按 seq 拼回来。
library;

import 'dart:convert';
import 'dart:typed_data';

/// 消息类型。改这里必须同步改手环端的 protocol.js。
class MsgType {
  MsgType._();

  static const String hello = 'hello';
  static const String schedule = 'schedule';
  static const String part = 'part';
  static const String config = 'config';
  static const String ready = 'ready';
  static const String request = 'request';
  static const String ack = 'ack';
}

/// 单帧最大字符数。与手环端 `createAssembler` 保持一致。
const int chunkChars = 300;

/// 打包一个信封。
String encodeMessage(String type, [Object? payload]) =>
    jsonEncode({'t': type, 'd': payload});

/// 拆开一个信封。解析失败返回 null。
class DecodedMessage {
  const DecodedMessage(this.type, this.data);

  final String type;
  final Object? data;

  @override
  String toString() => 'DecodedMessage($type)';
}

DecodedMessage? decodeMessage(String raw) {
  try {
    final parsed = jsonDecode(raw);
    if (parsed is! Map) return null;
    final type = parsed['t'];
    if (type is! String || type.isEmpty) return null;
    return DecodedMessage(type, parsed['d']);
  } catch (_) {
    return null;
  }
}

/// 把要发出去的文本转成线上字节：**UTF-8**。
///
/// ⚠️ 这里踩过坑：原来写的是 `Uint8List.fromList(text.codeUnits)` ——
/// `codeUnits` 给的是 UTF-16 码元，塞进 Uint8List 时每个码元被截成一个字节。
/// 纯 ASCII 看不出问题，但课表里有中文（课程名、教室），'高'(U+9AD8) 会变成
/// 单字节 0xD8，而正确的 UTF-8 是 `e9 ab 98`。手环端拿到的是**非法 UTF-8**，
/// 解出来是乱码 → `JSON.parse` 失败 → 静默丢弃，表现就是"发过去了、手环没反应"。
///
/// 对称地，接收方向也必须按 UTF-8 解（见 [decodePayloadBytes]）。
Uint8List encodePayloadBytes(String text) =>
    Uint8List.fromList(utf8.encode(text));

/// 把手环发来的字节按 UTF-8 解回文本。
///
/// 别用 `String.fromCharCodes(bytes)` —— 那是把每个字节当 UTF-16 码元，
/// 中文会变成乱码，和上面的坑是同一个。
String decodePayloadBytes(List<int> bytes) => utf8.decode(
      bytes,
      allowMalformed: true,
    );

/// 把整份课表 JSON 切成待发送的帧。
///
/// 课表足够小就只发一帧 `schedule`，否则发多帧 `part`（seq 从 0 起）。
/// 切片时不会把一个字符（代理对）从中间劈开 —— 手环端按字符拼接，
/// 劈开会产生非法 UTF-8。
List<String> buildFrames(String scheduleJson, {int chunk = chunkChars}) {
  if (scheduleJson.length <= chunk) {
    return [encodeMessage(MsgType.schedule, jsonDecode(scheduleJson))];
  }

  final pieces = <String>[];
  var i = 0;
  while (i < scheduleJson.length) {
    var end = i + chunk;
    if (end > scheduleJson.length) end = scheduleJson.length;
    // 不要把代理对劈成两半
    if (end < scheduleJson.length) {
      final unit = scheduleJson.codeUnitAt(end - 1);
      if (unit >= 0xD800 && unit <= 0xDBFF) end--;
    }
    pieces.add(scheduleJson.substring(i, end));
    i = end;
  }

  return [
    for (var seq = 0; seq < pieces.length; seq++)
      encodeMessage(MsgType.part, {
        'seq': seq,
        'total': pieces.length,
        'data': pieces[seq],
      }),
  ];
}

/// 手环回的 ack 内容。
class ScheduleAck {
  const ScheduleAck({required this.ok, this.courses = 0, this.error = ''});

  final bool ok;
  final int courses;
  final String error;

  static ScheduleAck? parse(Object? data) {
    if (data is! Map) return null;
    return ScheduleAck(
      ok: data['ok'] == true,
      courses: (data['courses'] as num?)?.toInt() ?? 0,
      error: (data['error'] as String?) ?? '',
    );
  }

  @override
  String toString() =>
      ok ? 'ack ok（手环显示 $courses 门课）' : 'ack 失败（$error）';
}
